### PR TITLE
Add an script to fork the hyperbox repositories

### PR DESCRIPTION
**Pull request by JaimeChavarriaga**
Add a bash script that configure and creates forks of the hyperbox repositories
Fix hyperbox/hyperbox#29

---
### Description
The root directory has a `fork` script to configure and create forks of the hyperbox repositories.
- The developer must clone the root repository and run the `init` script first.
- The script requires `curl` to create the fork-repositories in Github

---
### Running the script
The script configures an `upstream` remote with the original repository and change the `origin` repository to a fork-repository. It may configure the fork-repositories in the user account or in an organization. Optionally, it may create the forks in Github,

1. To configure the forks using repositories in a user account you must provide a *username*. F>or instance, if you provide the `user` as username, it configures fork-repositories such as https://github.com/user/hyperbox. The script does not create the forks if the `--create` option is not provided.
```
./fork user                      
```

2.  To configure the forks using an organization, you must use the `-o` or the `--org` options and provide an *organization* name. For instance, if you provide `organiz` as the *organization*, the script configures fork-repositories such as https://github.com/organiz/hyperbox.
``` 
./fork user --org organiz
```

3. To create the forks in Github, you must use the `-c` or the `--create` options. The script ask for the Github password and creates the forks. It supports passwords with white spaces and do not show the password in any `ps` listing. It uses `curl` and the Github REST API to create the forks.
``` 
./fork user --create
./fork user --org organiz --create
```

**NOTE:** You can run the script multiple times, for instance, if you wanna create new forks or had an error. The script shows error messages if the parameters are not provided correctly and if the cloned-repository has not been initialized.

### CODE REVIEW

File: fork [hunk 1]
```diff
@@ -0,0 +1,156 @@
+#!/bin/bash -e
+###########################################################################
+#
+# Hyperbox - Virtual Infrastructure Manager
+# Copyright (C) 2015 Maxime Dor
```
[the remark this answers is not in the document]
Comment (JaimeChavarriaga) [reply fork]: The header is based on the existing in the `init` script

File: fork [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+die() {
+    printf '%s\n' "$1" >&2
+    exit 1
+}
+
+# function to fork the repository in the current folder
+fork_current_repo () {
+    local PWD="$(pwd)"
+    local REPONAME="$(basename $PWD)"
+    local REPOURL="https://github.com/$ORGNAME/$REPONAME"
```
[the remark this answers is not in the document]
Comment (JaimeChavarriaga) [reply fork]: The problem here is that `git` does not support forks. Although it is possible to create the `origin` and the `upstream` remotes, it is not possible to create the fork-repository without using an specific API.
- Github has a `hub` command line tool that may use the git credentials and allows to fork the repository, but it is Github-specific.
- In Ubuntu, `hub` must be installed from a custom PPA. It is not included in the official repositories.

**Comment by JaimeChavarriaga** [comment]
The problem of just adding the additional remotes is the need to create the forks manually. Note that, by checking the `.gitmodules` information, it is possible to assure that all the git submodules are forked and people can create changes and open pull requests.

The updated `fork` script does not ask for the username. It uses the credentials from `git` and `hub`. As mentioned before, `git` cannot fork repositories in services such as Github. It is necessary to use provider-specific APIs or tools. By default, `hub` uses the git protocol instead HTTPS. However, it creates different remotes in the local git repository for each fork.

## Usage
To configure new remotes for the forks using the current user. 
```
$ ./fork
```
To create forks using the current user. 
```
$ ./fork --create
```

To configure new remotes the forks using an organization. 
```
$ ./fork --org organiz
```
To create the forks using an organization. 
```
$ ./fork --org organiz --create
```

To review the usage information
```
$ ./fork -h
```

## Requirements
This script only works with Github. It uses [the `hub` tool](https://github.com/github/hub). This tool can be installed in Ubuntu using the files in the release page of the Github repository

```
$ cd /temp
$ wget https://github.com/github/hub/releases/download/v2.5.1/hub-linux-amd64-2.5.1.tgz
$ tar xvzf hub-linux-amd64-2.5.1.tgz
$ cd hub-linux-amd64-2.5.1/
$ sudo ./install
```

**NOTE:** There is [a PPA to install the `hub` using `apt-get`](https://askubuntu.com/a/1023626). However, it installs an old version (the v.2.2.9) that does not support creating repositories in an organization, only under a username.

**Comment by maxidor** [comment]
So, having slept over this, my initial feedback stands: A fork script would only be accepted if it has high added value. Limiting it to Github doesn't add this high added value I'm afraid.

My personal idea is that this script makes little sense - a fork is not something you do everyday.
On the other hand, cloning might be done regularly enough that it warrants making the `init` script better to handle forks. I'm not recommending or asking you to do it. Only that if you want to pursue this, we need something that make sense, but so far I don't feel it does.

I want to thank you still for your PR and the work you've put into this, but it would be best if you contact us first before doing work that is about adding features/scripts, instead of bug fixing, as we might just not accept it.

I'll close this for now, but feel free to re-open if you want to continue and can reach something in line with the expectations outlined above.

**Pull request closed without merging**